crates/ruff_linter/src/rules/flake8_type_checking/helpers.rs

```diff
@@ -304,3 +297,91 @@ pub(crate) fn filter_contained(edits: Vec<Edit>) -> Vec<Edit> {
     }
     filtered
 }
+
+#[derive(Copy, PartialEq, Clone)]
+enum State {
+    Literal,
+    Annotated,
+    AnnotatedNonFirstElm,
+    Other,
+}
+
+pub(crate) struct QuoteAnnotation<'a> {
+    state: Vec<State>,
+    stylist: &'a Stylist<'a>,
+    annotation: String,
+    final_quote_type: Quote,
+}
+
+impl<'a> QuoteAnnotation<'a> {
+    pub(crate) fn new(stylist: &'a Stylist<'a>) -> Self {
+        let final_quote_type = stylist.quote();
+        Self {
+            state: vec![],
+            stylist,
+            annotation: String::new(),
+            final_quote_type,
+        }
+    }
+}
+
+impl<'a> source_order::SourceOrderVisitor<'a> for QuoteAnnotation<'a> {
+    fn visit_expr(&mut self, expr: &'a Expr) {
+        let generator = Generator::from(self.stylist);
+        match expr {
+            Expr::Subscript(ast::ExprSubscript { value, slice, .. }) => {
+                if let Some(name) = value.as_name_expr() {
+                    let value = generator.expr(value);
+                    self.annotation.push_str(&format!("{value}["));
```

Comment:
We should try to avoid doing any intermediate allocations
```suggestion
                    let value = generator.expr(value);
                    self.annotation.push_str(&value);
                    self.annotation.push_str("[");
```